shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/RoutePattern.kt

```diff
@@ -30,4 +32,109 @@ data class RoutePattern(
     fun isTypical() = typicality == null || typicality == Typicality.Typical
 
     override fun compareTo(other: RoutePattern) = sortOrder.compareTo(other.sortOrder)
+
+    data class PatternsForStop(
+        val allPatterns: List<RoutePattern>,
+        val patternsNotSeenAtEarlierStops: Set<String>
+    )
+
+    companion object {
+
+        /**
+         * Filter the given list of stopIds to the stops that don't have service redundant to
+         * earlier stops in the list; each stop must serve at least one route pattern that is not
+         * seen by any earlier stop.
+         */
+        fun filterStopsWithRedundantPatterns(
+            stopIds: List<String>,
+            globalData: GlobalResponse
+        ): List<String> {
+            val parentToAllStops = Stop.resolvedParentToAllStops(stopIds, globalData)
+
+            return patternsGroupedByLineOrRouteAndStop(
```

Comment:
I don't love that this is building data structures we don't care about, but seemed like the simplest approach. Since we run `patternsGroupedByLineOrRouteAndStop` every time we build route cards, I don't think this should have a negative performance impact.